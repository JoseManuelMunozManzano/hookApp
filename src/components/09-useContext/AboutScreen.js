// Para revisar información que otras páginas van a trabajar.
import React from 'react';

export const AboutScreen = () => {
  return (
    <div>
      <h1>AboutScreen</h1>
      <hr />
    </div>
  );
};
