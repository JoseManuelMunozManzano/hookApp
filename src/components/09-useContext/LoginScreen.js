// Aplicación que va a tener un login. Por ahora sólo se graba el usuario
import React from 'react';

export const LoginScreen = () => {
  return (
    <div>
      <h1>LoginScreen</h1>
      <hr />
    </div>
  );
};
