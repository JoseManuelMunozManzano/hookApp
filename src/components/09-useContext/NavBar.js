import React from 'react';
import { Link, NavLink } from 'react-router-dom';

// La diferencia entre el Link y el NavLink
// El NavLink da una característica extra, que es saber cual es la ruta en la cual me
// encuentro y activar alguna clase de CSS. Ver en el ejemplo activeClassName

export const NavBar = () => {
  return (
    // <nav>
    //   {/* Usar los anchor tags no es la manera correcta de usar la navegación
    //   cuando se esta trabajando con React Router.
    //   El problema es que se fuerza el refresco de la página, cosa que no queremos */}
    //   {/* No hay que usarlos. Hay que transformarlos en Links.  */}
    //   <ul>
    //     <li>
    //       {/* <a href="./">Home</a> */}
    //       <Link to="./">Home</Link>
    //     </li>
    //     <li>
    //       {/* <a href="./login">Login</a> */}
    //       <Link to="./login">Login</Link>
    //     </li>
    //     <li>
    //       {/* <a href="./about">About</a> */}
    //       <Link to="./about">About</Link>
    //     </li>
    //   </ul>
    // </nav>

    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <div className="container-fluid">
        <Link to="/" className="navbar-brand">
          useContext
        </Link>
        <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
          <div className="navbar-nav">
            {/* No olvidar el exact porque si no Home quedará siempre active */}
            <NavLink
              exact
              activeClassName="active"
              to="./"
              className="nav-link"
            >
              Home
            </NavLink>
            <NavLink
              exact
              activeClassName="active"
              to="./about"
              className="nav-link"
            >
              About
            </NavLink>
            <NavLink
              exact
              activeClassName="active"
              to="./login"
              className="nav-link"
            >
              Login
            </NavLink>
          </div>
        </div>
      </div>
    </nav>
  );
};
