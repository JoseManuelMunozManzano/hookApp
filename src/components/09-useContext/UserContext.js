import { createContext } from 'react';

// El UserContext es un Higher Order Component (como el Switch o el Router del fuente
// AppRouter.js)
// Esto significa que se le pueden colocar componentes hijos dentro de el.
export const UserContext = createContext(null);
