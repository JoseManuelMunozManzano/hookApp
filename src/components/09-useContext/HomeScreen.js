// Donde tenemos la información del usuario
import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const HomeScreen = () => {
  // Si aquí necesito acceso al usuario
  // Esto le dice al useContext que busque la siguiente instancia en el arbol de
  // componentes que sea del tipo UserContext
  // Aquí uso el user cuyo valor se ha dado en LoginScreen.js
  const { user } = useContext(UserContext);

  console.log(user);

  return (
    <div>
      <h1>HomeScreen</h1>
      <hr />

      <pre>{JSON.stringify(user, null, 3)}</pre>
    </div>
  );
};
