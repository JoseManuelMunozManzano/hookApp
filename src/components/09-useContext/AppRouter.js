// Fichero con las rutas de la aplicación
// Una app puede tener más de un router.
// El <div> dentro del <Router> se aconseja siempre crearlo. Sirve para contener toda
// la información que el router va a manejar, y puede haber más de uno. Se puede poner
// también un <Fragment> ó <>
// Para ver que ruta se usa, usamos el <Switch>
// La Ruta en si es el <Route> donde aparecerá el path (lo que aparece en el navegdor) y
// que componente renderiza. La mostrada en pantalla es una forma de hacerlo. En la
// documentación (https://reactrouter.com/web/guides/quick-start) aparece otra similar.
//
// Se sugiere colocar la ruta más general abajo, aunque no siempre va a ser el caso.
// Esto es porque si no se pone la propiedad exact, entonces el path no hace falta
// que sea el mismo, solo que contenga ese texto, y la ruta "/" contiene a todas las demás.
//
// Al poner exact={true} o exact solamente, estamos diciendo que el path tiene que ser
// exactamente ese.
//
// Los 3 componentes HomeScreen, LoginScreen y AboutScreen son como hermanos, están
// contenidos en el mismo nivel.
// Por tanto, si un componente generara información que necesito pasársela al HomeScreen,
// cómo lo hago? Con el context

import React from 'react';
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect,
} from 'react-router-dom';

import { NavBar } from './NavBar';

import { AboutScreen } from './AboutScreen';
import { HomeScreen } from './HomeScreen';
import { LoginScreen } from './LoginScreen';

export const AppRouter = () => {
  return (
    <Router>
      <div>
        <NavBar />

        <Switch>
          <Route exact path="/" component={HomeScreen} />

          <Route exact path="/about" component={AboutScreen} />
          <Route exact path="/login" component={LoginScreen} />

          {/* Si la ruta es erronea se puede: */}

          {/* 1. Hacer el default del switch */}
          {/* <Route component={HomeScreen} /> */}

          {/* 2. Hacer un redirect (recomendado porque la url del navegador queda bien) */}
          <Redirect to="/" />
        </Switch>
      </div>
    </Router>
  );
};
