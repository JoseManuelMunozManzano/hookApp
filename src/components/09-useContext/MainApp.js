// Aplicación como tal
import React from 'react';
import { AppRouter } from './AppRouter';
import { UserContext } from './UserContext';

// Si enmarco el AppRouter en el componente UserContext, significa que lo que se defina
// en UserContext va a estar disponible a lo largo de todos sus hijos (<AppRouter /> en
// este caso)
// Importante el .Provider. Sirve para proveer información a lo largo de todos sus
// componentes hijos. Lo que se desea compartir se debe hacer a través de una propiedad
// llamada value

export const MainApp = () => {
  const user = {
    id: 1234,
    name: 'José Manuel',
    email: 'xxxxxx@hotmail.es',
  };

  return (
    <UserContext.Provider value={user}>
      <AppRouter />
    </UserContext.Provider>
  );
};
