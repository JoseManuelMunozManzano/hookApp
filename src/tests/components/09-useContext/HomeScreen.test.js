import { mount } from 'enzyme';
import { HomeScreen } from '../../../components/09-useContext/HomeScreen';
import { UserContext } from '../../../components/09-useContext/UserContext';

describe('Pruebas en <HomeScreen />', () => {
  const user = {
    name: 'José Manuel',
    email: 'xxxx@gmail.com',
  };
  // Cómo se establece un contexto?
  // Se añade tal y como se hace en MainApp.js
  // Es doble llave en value porque es igual que en MainApp.js, estamos
  // devolviendo un objeto
  //
  // NOTA: el shallow sólo va a renderizar el primer componente, en este caso
  // UserContext, que lo único que hace es que internamente está renderizando
  // el HomeScreen. Por eso el snapshot generado tiene como resultado:
  // exports[`Pruebas en <HomeScreen /> debe de mostrarse correctamente 1`] = `<HomeScreen />`;
  const wrapper = mount(
    <UserContext.Provider value={{ user }}>
      <HomeScreen />
    </UserContext.Provider>
  );

  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
