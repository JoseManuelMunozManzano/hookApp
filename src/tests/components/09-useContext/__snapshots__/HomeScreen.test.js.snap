// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HomeScreen /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    HomeScreen
  </h1>
  <hr />
  <pre>
    {
   "name": "Jos√© Manuel",
   "email": "xxxx@gmail.com"
}
  </pre>
</div>
`;
